Replace `GeometricSpace` and `Cell` by `Representation`

ndaelman-hu opened the pull request:
This PR consolidates the cell representation architecture by replacing the deprecated `GeometricSpace`, `Cell`, and `AtomicCell` classes with a unified `Representation` + `RelativeRepresentation` aprpoach. While almost all quantities are retained, the structure now:

- Puts forward `ModelSystem` (root) as the default representation, whereas alternative representations are stored under `ModelSystem.representations`. This avoids unnecessary clutter for when there is only one representation.
- `RealtiveRepresentation` now acts as a base class for defining various kinds of base sections, not just crystallographic ones.

It also updates the type signature of `from_ase_atoms` and makes it a standalone method. This simplifies the initiation of `ModelSystem`s from `ase.Atoms` by providing a default clean starting point. The method';s mapping is also extended. These extensions only make sense in light of `Representation`

**NOTE**: it is strongly recommended to review this PR using _semantic diff_.

ndaelman-hu committed:
Add lattice_vectors, periodic_boundary_conditions, and fractional_coordinates directly under `ModelSystem`

ndaelman-hu committed:
Move AtomicCell out to ModelSystem

ndaelman-hu committed:
Move common properties to `Representation`

ndaelman-hu committed:
Move cell and geometric space under representation

ndaelman-hu committed:
Remove unused length and angle vector properties from GeometricSpace and Representation classes

- Removed length_vector_a/b/c and angle_vectors_b_c/a_c/a_b from both GeometricSpace and Representation classes
- These properties were automatically computed from lattice vectors and not essential data
- Updated assignment code to only set volume instead of length/angle measurements
- Simplifies schema by removing redundant computed properties that can be derived from lattice_vectors

ndaelman-hu committed:
- Move specialized quantities to `RelativeRepresenation`
- Restructure lay-out `Representation`
-- Make difference between hyper-volume and hyper-area more apparent

ndaelman-hu committed:
- Remove old sections GeometricSpace, Cell, AtomicCell
- Move out `from_ase_atoms` to produce components

ndaelman-hu committed:
Improve `from_ase_atoms`:
- Leverage `ModelSystem.normalize` to simplify type signature
- Explain architectural choices in pydocs string
- Extend mapping to include velocities, fractional_coordinates, charges, volumes, atom labels

ndaelman-hu committed:
Extend volume detection to other dimensionalities

ndaelman-hu committed:
- Fix tests
- Apply ruff

ndaelman-hu committed:
Fix mypy type annotations
  - Fix Optional parameter types in model_system.py (_copy_common_quantities)
  - Fix Optional parameter types in numerical_settings.py (None checks added)
  - Fix Optional parameter types in utils/utils.py (log decorator)
  - Replace AtomicCell with Representation in tests/utils/test_utils.py
  - Update test logic to use representations instead of cell subsection

ndaelman-hu commented:
@JFRudzinski I would also like to add GlobalSymmnetry here, from #253 
The reason is that it nicely shows how representations interact with symmetry.

ndaelman-hu committed:
- Fix CI/CD pytest errors
- Add backwards compatiblity layer for parsers

ndaelman-hu committed:
- Fix pytest errors
-- Patch `resolve_high_symmetry_points` and its tests

ladinesa commented:
@ndaelman-hu sorry i probably will not be able to look at this this week.

JFRudzinski commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -341,57 +252,72 @@ class AtomicCell(Cell):
-        atoms = parent.to_ase_atoms(logger=logger)
-        if atoms is None:
-            return  # parent already logged the problem
+class RelativeRepresentation(Representation):
```
Just a small comment (perhaps too nitpicky): Could this just be named `AlternateRepresentation` or `AlternativeRepresentation` or something like that? I feel that `Relative` puts too much focus on the transformation. More generally to symmetry transformations, it may not be of interest/scope/practicality to store the transformation function

JFRudzinski commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -256,83 +129,121 @@ def __eq__(self, other):
         description="""
-        If periodic boundary conditions are applied to each direction of the crystal axes.
+        Whether periodic boundary conditions are applied.
+        Runs over each direction of the crystal axes.
```
remove crystal specificity of description

JFRudzinski commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -256,83 +129,121 @@ def __eq__(self, other):
         shape=[3],
         description="""
-        If periodic boundary conditions are applied to each direction of the crystal axes.
+        Whether periodic boundary conditions are applied.
```
This is inherently linked to the lattice vectors, should we mention this? Should we add a warning if this is set yes without lattice vectors? (I guess only relevant if we have functionalities which apply the PBC)

ndaelman-hu replied:
I would rather leave it empty and log that.

JFRudzinski reviewed (approved):
Very nice! I like this new structure.

I see the back compatibility with cell, this is nice. Can you add an issue in the parser plugin branch to remove cell usage from all parsers, to make sure we tackle this at some point?

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -256,83 +129,121 @@ def __eq__(self, other):
 
 
-class Cell(GeometricSpace):
+class Representation(ArchiveSection):
```
Not sure why the need to define this base class separate from `RelativeRepresentation`? This is not used anywhere aside from cell backward compatibility, which I think we should not be doing at the development stage.

ndaelman-hu replied:
`Representation` is also a base class to `ModelSystem`, which is the default representation. It contains the common quantities definitions between `ModelSystem` and `ModelSystem.representations`.

It's not particularly meant for backwards compatibility, but may provide it (see bottom of `model_system.py`).

ndaelman-hu replied:
I will be resolving this convo, deciding to keep the class. If you want to discuss further, feel free to re-open.

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1027,12 +955,50 @@ class ModelSystem(System):
+
     sub_systems = SubSection(sub_section=SectionProxy('ModelSystem'), repeats=True)
 
+    # Backward compatibility: add cell as a subsection that references Representation
```
imo we do not need to provide this at the development stage it will just create a lot of clutter.

ndaelman-hu replied:
Sorry for the extraneous comment. No, `cell` is being removed as per the title of the PR.

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -766,7 +705,7 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
 
 
-class ModelSystem(System):
+class ModelSystem(System, Representation):
```
for sure I do not understand the logic here but in general multiple inheritances is a symptom of not optimal design

ndaelman-hu replied:
Yes, I'd prefer not using it either. Let's see:

- `System` comes from NOMAD, so the sub-optimal design are how definitions are split along the projects. 
- `Representation` is just a way to avoid defining the same quantities twice. I worst case, it can be removed, though I would explore other options first.
- Any specific tests I should run to see whether double inheritance causes an issue?

ndaelman-hu replied:
@ladinesa Can I resolve this?

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1027,12 +955,50 @@ class ModelSystem(System):
+        if atoms is None:
+            return  # parent already logged the problem
+
+        try:
```
The log decorator already does this, you can provide a custom warning via `exc_msg`

ndaelman-hu replied:
like this?

ladinesa replied:
something like
```
@log(exc_msg=''Failed to extract geometric-space data from ASE cell.'')
def get_geometric_space_for_cell
```
Though I have to update the log utl so it does not append the exact error but would add it as exc_info. Maybe you can include this tiny change in l58 of utils.py i.e. `_logger.warning(f'{_exc_msg}', exc_info=e)`

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1138,12 +1112,36 @@ def to_ase_atoms(self) -> 'ase.Atoms | None':
+        else:
+            # Use specified representation
+            if not self.representations:
+                raise ValueError('No representations available in ModelSystem.')
```
not sure how this is handled downstream but I would suggest not raising errors

ndaelman-hu replied:
Honestly, this is legacy stuff that has little usage in our schema. It just interfaces with `ase`...

ndaelman-hu replied:
swapped error raising for logging debug and info messages

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1155,9 +1153,9 @@ def to_ase_atoms(self) -> 'ase.Atoms | None':
-        else:
-            logger.warning('No cell section available in ModelSystem.')
+                logger.info(
+                    f'No lattice_vectors found in representations[{representation_index}].'
```
no formatted logs

ndaelman-hu replied:
done

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1565,3 +1538,155 @@ def is_molecule(self) -> bool:
+
+
+@log
+def from_ase_atoms(ase_atoms: ase.Atoms) -> ModelSystem:
```
this should at least be a static method.

ndaelman-hu replied:
I'm kind of torn here. For sure, I don't want this method operating directly on the `ModelSystem` object, as it introduces a whole host of unnecessary state checks. By instantiating from a new object we are sure to have a clean setup, that is simply filled up.

So, this method can stand by itself. Granted, given that `to_ase_atoms` is in the `ModelSystem` namespace, one could expect its dual within the same namespace.
`staticmethod` is indeed compatible with generating a new `ModelSystem` object, but from the interface alone, one would expect it to write to the object.

ladinesa replied:
I think a static method is really more apt for this.

ndaelman-hu replied:
Okay, it does not really clash with the clean state. It's more a question of managing expectations.

ladinesa commented on src/nomad_simulations/schema_packages/numerical_settings.py:
```diff
@@ -404,11 +411,14 @@ def get_k_line_density(
-            grid=self.grid,
+        if (
+            reciprocal_lattice_vectors is None
+            or not KSpaceFunctionalities().validate_reciprocal_lattice_vectors(
```
could be in a separate mr but looks like `validate_reciprocal_lattice_vectors` should be a static method

ndaelman-hu replied:
You're right. It's a small correction, I'll add it here. Thx for spotting this.

ladinesa commented on tests/conftest.py:
```diff
@@ -347,7 +350,7 @@ def model_system() -> ModelSystem:
 
 @pytest.fixture(scope='session')
-def atomic_cell() -> AtomicCell:
+def atomic_cell() -> Representation:
```
should you redefine fixture too atomic_representation?

ndaelman-hu replied:
done

ndaelman-hu commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1027,12 +955,50 @@ class ModelSystem(System):
 
+    # Backward compatibility: add cell as a subsection that references Representation
+    cell = SubSection(sub_section=Representation.m_def, repeats=True)
+
```
Added as a property after the `ModelSystem` definition

ndaelman-hu commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1027,12 +955,50 @@ class ModelSystem(System):
+                'Failed to extract geometric-space data from ASE cell.',
+                exc_info=exc,
+            )
+
```
@JFRudzinski do you think that we should retain quantities like vector length and angles? Note that they are also encoded in the lattice vectors. For cyrstals, the other symmetry descriptors also cover all the relevant angular information.

ndaelman-hu commented on tests/test_model_system.py:
```diff
@@ -218,14 +378,20 @@ def test_normalize(self):
+                if sys.symmetry.primitive_cell:
+                    assert sys.symmetry.primitive_cell.name == 'primitive'
+                if sys.symmetry.conventional_cell:
+                    assert sys.symmetry.conventional_cell.name == 'conventional'
```
done

ndaelman-hu committed:
Apply reviewer comments

ndaelman-hu committed:
Fix target for symmetry tests to `ModelSystem.representations`

ndaelman-hu committed:
- Fix mypy and pytest
- Apply ruff

ndaelman-hu committed:
Append the error to the structured logging

ladinesa commented on src/nomad_simulations/schema_packages/utils/utils.py:
```diff
@@ -56,7 +56,7 @@ def wrapper(*args, **kwargs):
                 )
             except Exception as e:
-                _logger.warning(f'{_exc_msg} {e}')
+                _logger.warning(f'{_exc_msg}', exc_info=e)
```
sorry you can undo this change as it appears it produced some unwanted failures in test, I can update this in a separate mr.

ndaelman-hu committed:
Revert "Append the error to the structured logging"

This reverts commit 3653c94c2ed7c087d5d2a293e944f03f5de30326.

ndaelman-hu committed:
- Improve generation of ChemicalSymbols directly via particle labels
-- TODO: stress test
- Incorporate representation setup into `generate_model_system`
- Remove defaults from `generate_model_system`, allowing for empty fields
- Fix `test_model_system.py/test_normalize`

ndaelman-hu committed:
Fix mypy error

ndaelman-hu committed:
Merge remote-tracking branch 'origin/develop' into add-representation-section

# Conflicts:
#	src/nomad_simulations/schema_packages/model_system.py

ndaelman-hu committed:
Fix undefined variable after merge: conventional_atomic_cell → conventional_cell

During the merge from origin/develop, the variable name was updated from
conventional_atomic_cell to conventional_cell to match the Representation
refactoring, but one reference on line 546 was missed.

ndaelman-hu committed:
Complete Representation refactoring: modernize tests and remove backward compatibility

- Remove AtomicCell and Cell class aliases
- Remove cell SubSection and property redirect from ModelSystem
- Update test_basis_set.py: cell=[AtomicCell()] → representations=[Representation()]
- Update test_general.py: .cell.append() → .representations.append()

Test results: 789 passed, 3 failed (99.6% pass rate)

ndaelman-hu committed:
Fix test helpers: remove dead 'type' parameter, add representation_name

- Remove unused 'type' parameter from generate_model_system()
- Add 'representation_name' parameter to properly name representations
- Update all callers: type='primitive' → representation_name='primitive'
- Fix test_normalize: add is_representative=True and correct parameters
- Fix parameter names: periodic_boundary_conditions → pbc, atomic_numbers → orbitals_symbols

All 792 tests now passing (100%)

ndaelman-hu committed:
Merge branch 'origin/develop' into add-representation-section

Resolved merge conflicts by integrating develop's new ElectronicState architecture with our Representation refactoring:

- model_system.py: Updated normalization documentation to match develop's clearer version
- conftest.py: Integrated develop's ElectronicState(basis_orbitals) pattern in generate_model_system(), replacing old OrbitalsState approach
- test_model_method.py: Activated develop's test cases (10) with Representation architecture (replaced AtomicCell references)

All tests passing (797/797).

ndaelman-hu committed:
Apply ruff

ndaelman-hu commented:
@JFRudzinski I merged in the latest updates from `develop`. I will do a quick double-check of the interface, but the first checks indicate a complete rollout. Should be merged soon.

ndaelman-hu commented:
So, one last point that keeps coming up: "encountered questions about how coordinate systems are handled across different frameworks". I also wanted to know whether our approach was consistent with community standards. This led to a comprehensive investigation of coordinate system conventions in computational materials science.

##  OPTIMADE API

In version 1.2.0 (the current stable release), OPTIMADE specifies only `cartesian_site_positions` for atomic positions, with no fractional coordinate support. The specification recommends that "the **first lattice vector SHOULD be set along x** and the second on the xy-plane" for databases that don't define an absolute Cartesian system (https://schemas.optimade.org/releases/v1.2.0/v1.2/properties/optimade/structures/lattice_vectors). This may reflect that different simulation codes may output structures with different orientations.

The development version (v1.3.0-rc.1), however, includes a new `fractional_site_positions` property alongside the existing Cartesian positions (https://github.com/Materials-Consortia/OPTIMADE/blob/develop/optimade.rst), similar to NOMAD's dual storage approach.

 ## ASE (Atomic Simulation Environment)

[ASE](https://ase-lib.org/ase/cell.html) stores cell vectors as row vectors in a 3×3 matrix, where `cell[i, j]` represents the jth Cartesian coordinate of the ith lattice vector. Positions are stored in Cartesian coordinates by default, with fractional coordinates computed on-demand via `scaled_positions()` and `cartesian_positions()` methods. Critically, ASE makes no specification about the absolute orientation of the Cartesian frame—it's **implicit and determined by how the structure was defined**. This matches NOMAD's approach exactly, which is why conversion between ASE and NOMAD is trivial with no coordinate transformations needed.

##  Spglib and the Transformation Matrix

This package is the origin of `transformation_matrix` and `origin_shift` (had to track back to git commit 8ccf1e6). These properties were introduced "to define the conventional cell" and appeared in **two different classes** with different documentation. In `GeometricSpace` (now `Representation`), the language about "custom coordinates system" vs "global coordinates system" is unclear. It suggests that any modification should refer to the "representative" system's frame. Meanwhile, in `GlobalCrystalSymmetry` they were clearly described as relating "the origin of `ModelSystem.cell` to the conventional unit cell that was used to **derive the space group symmetry**." The latter should be reserved for the `Symmetry` section.

The actual implementation showed these values come directly from spglib's `_get_spglib_origin_shift()` and `_get_spglib_transformation_matrix()` methods. Spglib's documentation
  (https://spglib.readthedocs.io/en/latest/definition.html and https://spglib.readthedocs.io/en/stable/dataset.html) interpretes **these transformations operate entirely in fractional coordinate space, not Cartesian space**. As spglib's documentation explicitly states, "the transformation matrix doesn't rotate a crystal in Cartesian coordinates, but just changes the choices of basis vectors."

Spglib thus takes an input cell and produces a standardized conventional cell that matches crystallographic conventions for the identified space group. The transformation relates **how fractional coordinates map between these cells via x_std = P x_input + p, where both cells are expressed in the same implicit Cartesian frame**. The lattice vectors transform as (a_std, b_std, c_std) = 
  (a_input, b_input, c_input) P^-1. Crucially, **Cartesian positions of atoms don't change**—they're just described with different fractional coordinates relative to different lattice vectors.

##  DFT Simulation Codes

Major DFT codes (e.g. VASP, Quantum ESPRESSO, CASTEP) handle coordinate input. All three codes accept both Cartesian and fractional coordinates in their input files. While the materials modeling community (https://mattermodeling.stackexchange.com/questions/1063) agrees that input format doesn't affect performance, fractional coordinates are strongly recommended for periodic systems because they better preserve crystallographic symmetries during geometry optimization. These symmetries are crucial for computational efficiency.

Internally, different codes likely use different representations optimized for their algorithms, but the conversion cost between representations is negligible. The takeaway is that **the community standard is to work with fractional coordinates for periodic systems** when possible, but codes are flexible about input format.

@EBB2675 explained me that in most molecular codes, the basis is also implicit, and rather arbitrary. Even when alignment is highly relevant, e.g. spectroscopy of polarized light, the code will enforce the orientation of the molecules via their positions rather than adjusting some reference frame. I had a similar experience with Gaussian regarding reference frames. Still, we decided that it would be good to add a comment for cases where the reference frame and crystal lattice diverge, or only the former is defined. This can then be added when we encounter a parsing example. 

##  Implications for NOMAD

This investigation validated the new "implicit Cartesian frame" approach that I defined here. All major frameworks (ASE, OPTIMADE, spglib) use the same model: an implicit Cartesian reference frame that isn't stored explicitly, with both lattice vectors and positions expressed in this frame. The confusion in our documentation arose from inheriting spglib's terminology about "global" vs "custom"  coordinate systems, when in reality both systems share the same Cartesian basis and differ only in their choice of lattice vectors.

Our documentation updates

- clarify that the Cartesian frame is implicit and conventional rather than stored
- simplify mathematical descriptions to avoid overwhelming users
- explain that `transformation_matrix` and `origin_shift` relate lattice vectors within the same Cartesian frame rather than transforming between different Cartesian bases

JFRudzinski commented:
Thank you for the summary @ndaelman-hu This all looks reasonable to me :+1:

ndaelman-hu committed:
- Implement compute_fractional_coordinates() + tests
  - Add TODO for future 2D/1D system support
- Move positions quantity from Representation to ModelSystem to enforce fixed positions across representations
- Improve documentation for Representation and ModelSystem classes:
  - Clarify implicit Cartesian coordinate system convention
  - Document lattice_vectors as row-major 3x3 matrix
  - Add links to comprehensive architecture documentation
  - Clarify transformation_matrix and origin_shift semantics
  - Update examples to reflect Representation architecture changes

Changes to be committed:
	modified:   src/nomad_simulations/schema_packages/model_system.py
	modified:   tests/test_model_system.py

ndaelman-hu committed:
Add comprehensive documentation for Representation architecture:
  - Replace stub model_system.md with complete guide covering unified architecture, direct property access, alternative representations, normalization workflow, and quick start examples
  - Add representation.md documenting geometric representation system, Cartesian coordinate conventions, and integration with symmetry analysis
  - Add representation-squash-commit.md capturing design history, migration guide from old Cell/AtomicCell architecture, and design rationale
  - Restructure mkdocs.yml navigation to improve documentation discoverability with organized subsections for ModelSystem, AtomsState, and ModelMethod

Changes to be committed:
	modified:   docs/model_system/model_system.md
	new file:   docs/model_system/representation-squash-commit.md
	new file:   docs/model_system/representation.md
	modified:   mkdocs.yml
